File: concepts/07_modules_packages_and_loading/01_modules_imports_linkage_and_live_bindings/lua/test_02_binding_aliases_and_mutable_exports.lua
-- Common question: do imported aliases remain live when exports or local bindings change?
-- Inputs: a cached module table, copied scalar field, table alias, and package.loaded replacement.
-- Observations: alias identity, scalar snapshots, field mutation visibility, and future-require behavior.
-- polyglot-family: modules_packages_and_loading
-- polyglot-concept: modules_imports_linkage_and_live_bindings
-- polyglot-related: languages/lua/tooling_and_runtime/13_modules_and_loading/test_099_require_cache_and_lua_modules.lua

local t = require("support.assertions")
package.loaded.course_sample = nil
local exports = require("course_sample")
local alias = exports
local copied_answer = exports.answer

exports.answer = 43
t.equal(alias.answer, 43)
t.equal(copied_answer, 42)

local replacement = {answer = 99}
package.loaded.course_sample = replacement
t.same(require("course_sample"), replacement)
t.same(alias, exports)
t.equal(alias.answer, 43)

t.done()

File: concepts/07_modules_packages_and_loading/01_modules_imports_linkage_and_live_bindings/ruby/test_02_binding_aliases_and_mutable_exports.rb
# 共同问题：导入 alias 在导出状态改变后仍是 live binding 还是快照。
# 输入：Module alias、标量读取、可变 Array 和 writer；观察：常量 alias 身份、标量快照及共享导出。
# polyglot-family: modules_packages_and_loading
# polyglot-concept: modules_imports_linkage_and_live_bindings
# polyglot-related: languages/ruby/language/04_classes_modules_and_lookup/
# polyglot-related+: test_032_class_variables_class_state_and_constants.rb

require "assertions"

A = PolyglotAssertions

exports = Module.new do
  @value = 1
  @items = []

  class << self
    attr_accessor :value
    attr_reader :items
  end
end

alias_value = exports
scalar_snapshot = exports.value
items_alias = exports.items
exports.value = 2
exports.items << :ruby

A.same(exports, alias_value)
A.equal(1, scalar_snapshot)
A.equal(2, alias_value.value)
A.same(items_alias, exports.items)
A.equal([:ruby], items_alias)

A.done
